substrate/frame/contracts/proc-macro/src/lib.rs

```diff
@@ -217,6 +224,16 @@ impl HostFn {
 				id => return Err(err(span, &format!("Unsupported attribute \"{id}\". {msg}"))),
 			}
 		}
+
+		if mutating {
+			let stmt = syn::parse_quote! {
+				if ctx.ext().is_read_only() {
+					return Err(Error::<E::T>::StateChangeDenied.into());
+				}
+			};
+			item.block.stmts.insert(0, stmt);
+		}
```

Comment:
As stated before: We can't return that early here. This codes runs before we charge the gas for the host function call invocation itself. This has to be charged even if no worked has to be done. You have to move down this statement after:
https://github.com/paritytech/polkadot-sdk/blob/89604daa0f4244bc83782bd489918cfecb81a7d0/substrate/frame/contracts/proc-macro/src/lib.rs#L708-L711

Reply:
Ahh yes. Can you add a test that makes sure this weight is charged when erroring out early?

Reply:
I tried to test it using fixture:
```
pub extern "C" fn call() {
	input!(n: u32, );

	match n {
		0 => unsafe {core::ptr::read_volatile(core::ptr::null())},
		_ => api::set_storage(&[1u8; 32], &[1u8; 1]),
	}
}
```
The fixture is called with n equal to 0 and 1. Both calls result in a trap, but the difference is more significant than just the host function call cost because the error handling differs for both cases (WASMI execution time). The difference is the host function call cost plus 10%.

Reply:
You can call the `tokens` function on the gas meter to check which tokens were charged.

Reply:
As discussed, we do not have access to gas_meter in our tests. We could add such functionality to our tests to trace gas metering more deeply during testing. This could be implemented as a separate PR.